### PR TITLE
Bug fixes and general improvements

### PR DESCRIPTION
**Pull request by kleisauke**
Source code:
- Fix `attempt to index global 'vobject' (a nil value)` when setting an array image GValue (unit test included).
https://github.com/jcupitt/lua-vips/blob/3437523161702545963c8b6d39a3727ee8061f92/src/vips/gvalue.lua#L328
- Fix `attempt to index global 'vips' (a nil value)` when an error was thrown (unit test included).
https://github.com/jcupitt/lua-vips/blob/2d951e1ea87b26413381c16c4388a9596c1cb95d/src/vips/verror.lua#L18
- Fix missing options to `voperation.call` (for `bandjoin` and `bandrank`).
- Fix many white-space and other linter issues. `luacheck .` does now pass cleanly.
- Expose `Image.find_load` and `Image.find_load_buffer` (use it to work out what options to pass to `new_from_file` and `new_from_buffer`, I was missing it).

Unit tests:
- Pre-load the FFI module so that we can use `busted .` for the unit tests.
- Move `almost_equal` assertion to the buster helper.
- Test the missing GValue setters and getters.
- Test errors when loading from unknown buffer or when a file does not exits.
- Move initialization logic to `setup()`.
- gitignore `luacov.stats.out` and `luacov.report.out`.

### CODE REVIEW

File: spec/helpers.lua [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+local assert = require "luassert.assert"
+local say = require "say"
+
+-- TODO: Wait for https://github.com/Olivine-Labs/luassert/issues/148
+-- Meanwhile, we're applying patch #150 here.
+
+-- Pre-load the ffi module, such that it becomes part of the environment
+-- and Busted will not try to GC and reload it. The ffi is not suited
+-- for that and will occasionally segfault if done so.
+local ffi = require "ffi"
```
Comment (jcupitt) [inline spec/helpers.lua:13]: I like this fix for busted, nice! I never figured this one out.

File: spec/luacov.stats.out [hunk 2]
```diff
@@ -1,54 +0,0 @@
-163:/home/john/.luarocks/share/lua/5.1/busted/block.lua
```
Comment (jcupitt) [inline spec/luacov.stats.out:1]: I think this file should not be here -- you have stats.out in the .gitignore.

File: .gitignore [hunk 3]
```diff
@@ -1 +1,5 @@
 .*.swp
+
+# LuaCov
+luacov.stats.out
+luacov.report.out
```
Comment (jcupitt) [inline .gitignore]: is there a missing end of line?

File: .luacheckrc [hunk 4]
```diff
@@ -0,0 +1,3 @@
+files["spec/**/*.lua"] = {
+    std = "+busted",
+}
```
Comment (jcupitt) [inline .luacheckrc]: missing EOL again

File: src/vips/gvalue.lua [hunk 5]
```diff
@@ -320,12 +320,12 @@ local gvalue_mt = {
...
                 result = {}
                 for i = 0, pint[0] - 1 do
-                    -- this will make a new cdata object 
+                    -- this will make a new cdata object
                     local vimage = array[i]
 
                     -- vips_value_get_array_image() adds a ref for each image in
                     -- the array ... we must remember to drop them
-                    vobject.new(vimage)
+                    gobject_lib.g_object_ref(vimage)
```
Comment (jcupitt) [inline src/vips/gvalue.lua:328]: Can you explain this change? It looks like a ref leak to me.

File: src/vips/voperation.lua [hunk 6]
```diff
@@ -167,58 +163,56 @@ local voperation_mt = {
...
-                call_args
-            )
+            local match_image = find_order(function(x)
+                if Image.is_Image(x) then
+                    return x
+                else
+                    return nil
+                end
+            end,
+                call_args)
```
Comment (jcupitt) [inline src/vips/voperation.lua]: This indenting looks very odd! But I guess if it's what the linter likes. 

Perhaps we should swap the args to find_order()?

**Comment by jcupitt** [comment]
This looks great Kleis!

File: spec/helpers.lua (hunk 1 above)
Comment (kleisauke) [reply spec/helpers.lua:13]: Credits to: https://github.com/Olivine-Labs/luassert/pull/150

File: spec/luacov.stats.out (hunk 2 above)
Comment (kleisauke) [reply spec/luacov.stats.out:1]: It's already deleted, see: https://github.com/kleisauke/lua-vips/tree/cleanup/spec

File: .gitignore (hunk 3 above)
Comment (kleisauke) [reply .gitignore]: Yes nice catch! Fixed with: https://github.com/kleisauke/lua-vips/commit/b8917b960ad44f1d325d946caedbd47f44ca94c7

File: .luacheckrc (hunk 4 above)
Comment (kleisauke) [reply .luacheckrc]: Arghh, fixed with: https://github.com/kleisauke/lua-vips/commit/b8917b960ad44f1d325d946caedbd47f44ca94c7

File: src/vips/voperation.lua (hunk 6 above)
Comment (kleisauke) [reply src/vips/voperation.lua]: Hopefully fixed with: https://github.com/kleisauke/lua-vips/commit/b8917b960ad44f1d325d946caedbd47f44ca94c7. The auto-reformat command from my IDE does sometimes strange things..

File: src/vips/gvalue.lua (hunk 5 above)
Comment (kleisauke) [reply src/vips/gvalue.lua:328]: All other language bindings (except node-vips(?)) did this too, so I assumed it was the correct way. Without this change https://github.com/kleisauke/lua-vips/blob/cleanup/spec/gvalue_spec.lua#L71-L76 is failing.

node-vips: https://github.com/jcupitt/node-vips/blob/master/lib/gvalue.js#L281
ruby-vips: https://github.com/jcupitt/ruby-vips/blob/master/lib/vips/gvalue.rb#L209
pyvips: https://github.com/jcupitt/pyvips/blob/master/pyvips/gvalue.py#L301
net-vips: https://github.com/kleisauke/net-vips/blob/master/src/NetVips/GValue.cs#L479

Comment (jcupitt) [reply src/vips/gvalue.lua:328]: Oh huh then I'm sure you're right. Let's merge and I'll double-check.

**Comment by jcupitt** [comment]
Great work!

How's that rewrite going? I'd be curious to see any benchmarks.

**Comment by jcupitt** [comment]
Sorry, I forgot to ask, would you mind writing something for the ChangeLog?

**Comment by kleisauke** [comment]
Rewrite is going great. All the image manipulations have already been rewritten. You can follow the process here: https://github.com/weserv/images/tree/4.x.

I'm still working on the unit tests (aiming for 100% test coverage, so it takes a bit longer). 

I've not really started benchmarking but I can notice a speed increase. I've set-up a test environment here:
API 3 (php-vips):
https://t0.nl/
php-vips

API 4 (lua-vips):
https://t0.nl/?api=4
lua-vips

This is without caching, so it'll load a bit slower. I've also hard-coded the index images to load from file instead than downloading it from the origin url. The green-bar is the most important (TTFB).

OpenResty configuration still needs to be fine-tuned, so it might go faster later (need to experiment with LuaJIT settings and `init_by_lua_block` etc.). Also note that this is not our production server. The production one has more ram, more computing capacity and a fine-tuned kernel to handle more connections (on peak, we resize 4 million (4×10<sup>6</sup>) images per hour).

**Comment by kleisauke** [comment]
Changelog update can be found here: https://github.com/jcupitt/lua-vips/pull/23.

**Comment by jcupitt** [comment]
Yes, it looks quite a bit better already. Nice!

**Pull request merged**